src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs

```diff
@@ -74,18 +74,114 @@ private BoundNode VisitBinaryOperatorBase(BoundBinaryOperatorBase binaryOperator
 
         private T GetUpdatedSymbol<T>(BoundNode expr, T sym) where T : Symbol?
         {
-            if (sym is null) return sym;
+            if (sym is null || _snapshotManager is null) return sym;
 
-            if (_updatedSymbols.TryGetValue((expr, sym), out var updatedSymbol))
+            if (!_snapshotManager.TryGetUpdatedSymbol(expr, sym, out var updatedSymbol))
             {
-                Debug.Assert(updatedSymbol is object);
-                return (T)updatedSymbol;
+                updatedSymbol = sym;
+            }
+            Debug.Assert(updatedSymbol is object);
+
+            switch (updatedSymbol)
+            {
+                case LambdaSymbol lambda:
+                    return (T)remapLambda((BoundLambda)expr, lambda);
+
+                case LocalFunctionSymbol localFunction:
+                    return (T)remapLocalFunction(localFunction);
+
+                case LocalSymbol local:
+                    return (T)remapLocal(local);
+
+                case ParameterSymbol param:
+                    if (_remappedSymbols.TryGetValue(param, out var updatedParam))
+                    {
+                        return (T)updatedParam;
+                    }
+                    break;
             }
 
-            return sym;
+            return (T)updatedSymbol;
+
+            Symbol remapLambda(BoundLambda boundLambda, LambdaSymbol lambda)
+            {
+                var updatedDelegateType = _snapshotManager!.GetUpdatedDelegateTypeForLambda(lambda);
+
+                if (!_remappedSymbols.TryGetValue(lambda.ContainingSymbol, out Symbol? updatedContaining) && updatedDelegateType is null)
+                {
+                    return lambda;
+                }
+
+                LambdaSymbol updatedLambda;
+                if (updatedDelegateType is null)
+                {
+                    Debug.Assert(updatedContaining is object);
+                    var parameterRefKinds = lambda.ParameterRefKinds.IsDefault ? ImmutableArray<RefKind>.Empty : lambda.ParameterRefKinds;
```

Comment:
>var parameterRefKinds = lambda.ParameterRefKinds.IsDefault ? ImmutableArray<RefKind>.Empty : lambda.ParameterRefKinds; [](start = 20, length = 118)

It feels strange that CreateLambdaSymbol doesn't just take ParameterRefKinds from the original lambda and requires them to be adjusted in some way. If that is really needed, It would probably make sense to encapsulate the adjustment inside the API.